kotlinx-coroutines-core/common/src/channels/BroadcastChannel.kt

```diff
@@ -70,6 +74,333 @@ public fun <E> BroadcastChannel(capacity: Int): BroadcastChannel<E> =
+ * This channel is also created by `BroadcastChannel(Channel.CONFLATED)` factory function invocation.
+ *
+ * In this implementation, [opening][openSubscription] and [closing][ReceiveChannel.cancel] subscription
+ * takes linear time in the number of subscribers.
+ *
+ * **Note: This API is obsolete since 1.5.0.** It will be deprecated with warning in 1.7.0
+ * and with error in 1.8.0. It is replaced with [StateFlow][kotlinx.coroutines.flow.StateFlow].
+ */
+@Suppress("INVISIBLE_REFERENCE", "INVISIBLE_MEMBER")
+@ObsoleteCoroutinesApi
+public class ConflatedBroadcastChannel<E> private constructor(private val broadcast: BroadcastChannelImpl<E>)
+    : BroadcastChannel<E> by broadcast
+{
+    public constructor(): this(BroadcastChannelImpl<E>(capacity = CONFLATED))
+    /**
+     * Creates an instance of this class that already holds a value.
+     *
+     * It is as a shortcut to creating an instance with a default constructor and
+     * immediately sending an element: `ConflatedBroadcastChannel().apply { offer(value) }`.
+     */
+    public constructor(value: E) : this() {
+        trySend(value)
+    }
+
+    /**
+     * The most recently sent element to this channel.
+     *
+     * Access to this property throws [IllegalStateException] when this class is constructed without
+     * initial value and no value was sent yet or if it was [closed][close] without a cause.
+     * It throws the original [close][SendChannel.close] cause exception if the channel has _failed_.
+     */
+    public val value: E get() = broadcast.value
+    /**
+     * The most recently sent element to this channel or `null` when this class is constructed without
+     * initial value and no value was sent yet or if it was [closed][close].
+     */
+    public val valueOrNull: E? get() = broadcast.valueOrNull
+}
+
+/**
+ * A common implementation for both the broadcast channel with a buffer of fixed [capacity]
+ * and the conflated broadcast channel (see [ConflatedBroadcastChannel]).
+ *
+ * **Note**, that elements that are sent to this channel while there are no
+ * [openSubscription] subscribers are immediately lost.
+ *
+ * This channel is created by `BroadcastChannel(capacity)` factory function invocation.
+ */
+internal open class BroadcastChannelImpl<E>(
+    /**
+     * Buffer capacity; [Channel.CONFLATED] when this broadcast is conflated.
+     */
+    val capacity: Int
+) : BufferedChannel<E>(capacity = Channel.RENDEZVOUS, onUndeliveredElement = null), BroadcastChannel<E> {
+    init {
+        require(capacity >= 1 || capacity == CONFLATED) { "ArrayBroadcastChannel capacity must be at least 1, but $capacity was specified" }
+    }
+
+    // All operations are protected by this lock.
+    private val lock = ReentrantLock()
+    // The list of subscribers; all accesses should be protected by lock.
+    private val subscribers: MutableList<BufferedChannel<E>> = mutableListOf()
```

Comment:
CoW lists are used in subscription management for a reason.

Consider the following test:
```
    object ImmediateDispatcher : CoroutineDispatcher() {
        override fun dispatch(context: CoroutineContext, block: Runnable) {
            block.run()
        }
    }

    @Test
    fun testBreakBroadcastChannel() = runTest {
        val broadcast = BroadcastChannel<Int>(1)
        var r2: Job? = null
        val r1 = launch(ImmediateDispatcher) {
            try {
                broadcast.openSubscription().receive()
            } catch (e: Throwable) {
                r2!!.cancel()
            }
        }

        r2 = launch(ImmediateDispatcher) {
            val s = broadcast.openSubscription()
            try {
                s.receive()
            } catch (e: Throwable) {
                s.cancel()
            }
        }

        launch(ImmediateDispatcher) {
            try {
                broadcast.openSubscription().receive()
            } catch (e: Throwable) {
              //
            }
        }


        yield()
        broadcast.close()
    }
```

It, unexpectedly, fails with `ConcurrentModificationException`. Such dispatcher is not presented in out-of-the-box coroutines, but there are at least few 3rd-party code that uses it that I'm aware of.

Broadcast channels are obsolete, so no hard feelings about their performance, it should be sufficient to take a snapshot of `subscribers` for each meaningful operation